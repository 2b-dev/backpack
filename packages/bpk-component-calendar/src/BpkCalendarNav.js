import React, { PropTypes } from 'react';

import ArrowLeftIcon from 'bpk-component-icon/sm/arrow-left';
import ArrowRightIcon from 'bpk-component-icon/sm/arrow-right';
import BpkSelect from 'bpk-component-select';

import addMonths from 'date-fns/add_months';
import { getMonthsInRange, isWithinRange, getMonthRange, formatIsoMonth, parseIsoDate, startOfMonth } from './utils';

import './bpk-calendar.scss';

const BpkCalendarNav = (props) => {
  const {
    id,
    month,
    minDate,
    maxDate,
    formatMonth,
    onChangeMonth,
    changeMonthLabel,
  } = props;

  const baseMonth = startOfMonth(month);
  const { min, max } = getMonthRange(minDate, maxDate);
  const navigatableMonths = getMonthsInRange(minDate, maxDate);
  const prevMonth = addMonths(baseMonth, -1);
  const nextMonth = addMonths(baseMonth, 1);

  // Safeguard for disabled buttons is due to React bug in Chrome: https://github.com/facebook/react/issues/8308
  // PR: https://github.com/facebook/react/pull/8329 - unresolved as of 22/12/2016
  return (
    <div className="bpk-calendar-nav">
      <div className="bpk-calendar-nav__nudger">
        <button
          type="button"
          className="bpk-calendar-nav__button"
          onClick={() => {
            if (isWithinRange(prevMonth, min, max)) {
              onChangeMonth(prevMonth);
            }
          }}
          disabled={!isWithinRange(prevMonth, min, max)}
        >
          <ArrowLeftIcon className="bpk-calendar-nav__icon" />
          <span className="visually-hidden">{ formatMonth(addMonths(baseMonth, -1)) }</span>
        </button>
      </div>
      <div className="bpk-calendar-nav__month">
        <label htmlFor={`${id}_select`} className="visually-hidden">{ changeMonthLabel }</label>
        <BpkSelect
          id={`${id}_select`}
          name="months"
          value={formatIsoMonth(baseMonth)}
          onChange={event => onChangeMonth(parseIsoDate(event.target.value))}
        >
          { navigatableMonths.map(m => (
            <option value={formatIsoMonth(m)} key={m.toString()}>{ formatMonth(m) }</option>
          ))}
        </BpkSelect>
      </div>
      <div className="bpk-calendar-nav__nudger">
        <button
          type="button"
          className="bpk-calendar-nav__button"
          onClick={() => {
            if (isWithinRange(nextMonth, min, max)) {
              onChangeMonth(nextMonth);
            }
          }}
          disabled={!isWithinRange(addMonths(baseMonth, 1), min, max)}
        >
          <ArrowRightIcon className="bpk-calendar-nav__icon" />
          <span className="visually-hidden">{ formatMonth(addMonths(baseMonth, 1)) }</span>
        </button>
      </div>
    </div>
  );
};

BpkCalendarNav.propTypes = {
  id: PropTypes.string.isRequired,
  changeMonthLabel: PropTypes.string.isRequired,
  month: PropTypes.instanceOf(Date).isRequired,
  minDate: PropTypes.instanceOf(Date).isRequired,
  maxDate: PropTypes.instanceOf(Date).isRequired,
  formatMonth: PropTypes.func.isRequired,
  onChangeMonth: PropTypes.func.isRequired,
};

export default BpkCalendarNav;
