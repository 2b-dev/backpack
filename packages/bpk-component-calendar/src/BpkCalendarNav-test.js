import React from 'react';
import renderer from 'react-test-renderer';
import format from 'date-fns/format';
import addMonths from 'date-fns/add_months';

import BpkCalendarNav from './BpkCalendarNav';

const formatMonth = date => format(date, 'MMMM YYYY');

describe('BpkCalendarNav', () => {
  it('should render correctly', () => {
    const tree = renderer.create(<BpkCalendarNav
      month={new Date()}
      formatMonth={formatMonth}
      id="myCalendarNav"
    />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly with "minDate" and "maxDate" set', () => {
    const tree = renderer.create(<BpkCalendarNav
      month={new Date()}
      formatMonth={formatMonth}
      minDate={new Date()}
      maxDate={addMonths(new Date(), 2)}
      id="myCalendarNav"
    />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
