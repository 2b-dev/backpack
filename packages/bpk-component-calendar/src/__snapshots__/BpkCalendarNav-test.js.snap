exports[`BpkCalendarNav should render correctly 1`] = `
<div
  className="bpk-calendar-nav">
  <div
    className="bpk-calendar-nav__nudger">
    <button
      className="bpk-calendar-nav__button"
      disabled={true}
      onClick={[Function onClick]}>
      <svg
        className="bpk-calendar-nav__icon"
        height="18"
        viewBox="0 0 18 18"
        width="18"
        xmlns="http://www.w3.org/2000/svg">
        <path
          d="M14 2.5v13.1c0 1.3-1.5 1.9-2.4 1.1l-7.1-6.6c-.7-.6-.7-1.6 0-2.2l7.1-6.5c.9-.9 2.4-.2 2.4 1.1z" />
      </svg>
      <span
        className="visually-hidden">
        November 2016
      </span>
    </button>
  </div>
  <div
    className="bpk-calendar-nav__month">
    <select
      className="bpk-select"
      id="myCalendarNav_select"
      name="months"
      onChange={[Function onChange]}
      value="2016-12" />
  </div>
  <div
    className="bpk-calendar-nav__nudger">
    <button
      className="bpk-calendar-nav__button"
      disabled={true}
      onClick={[Function onClick]}>
      <svg
        className="bpk-calendar-nav__icon"
        height="18"
        viewBox="0 0 18 18"
        width="18"
        xmlns="http://www.w3.org/2000/svg">
        <path
          d="M4 2.5v13.1c0 1.3 1.5 1.9 2.4 1.1l7.1-6.5c.7-.6.7-1.6 0-2.2L6.4 1.4C5.5.5 4 1.2 4 2.5z" />
      </svg>
      <span
        className="visually-hidden">
        January 2017
      </span>
    </button>
  </div>
</div>
`;

exports[`BpkCalendarNav should render correctly with "minDate" and "maxDate" set 1`] = `
<div
  className="bpk-calendar-nav">
  <div
    className="bpk-calendar-nav__nudger">
    <button
      className="bpk-calendar-nav__button"
      disabled={true}
      onClick={[Function onClick]}>
      <svg
        className="bpk-calendar-nav__icon"
        height="18"
        viewBox="0 0 18 18"
        width="18"
        xmlns="http://www.w3.org/2000/svg">
        <path
          d="M14 2.5v13.1c0 1.3-1.5 1.9-2.4 1.1l-7.1-6.6c-.7-.6-.7-1.6 0-2.2l7.1-6.5c.9-.9 2.4-.2 2.4 1.1z" />
      </svg>
      <span
        className="visually-hidden">
        November 2016
      </span>
    </button>
  </div>
  <div
    className="bpk-calendar-nav__month">
    <select
      className="bpk-select"
      id="myCalendarNav_select"
      name="months"
      onChange={[Function onChange]}
      value="2016-12">
      <option
        value="2016-12">
        December 2016
      </option>
      <option
        value="2017-01">
        January 2017
      </option>
      <option
        value="2017-02">
        February 2017
      </option>
    </select>
  </div>
  <div
    className="bpk-calendar-nav__nudger">
    <button
      className="bpk-calendar-nav__button"
      disabled={false}
      onClick={[Function onClick]}>
      <svg
        className="bpk-calendar-nav__icon"
        height="18"
        viewBox="0 0 18 18"
        width="18"
        xmlns="http://www.w3.org/2000/svg">
        <path
          d="M4 2.5v13.1c0 1.3 1.5 1.9 2.4 1.1l7.1-6.5c.7-.6.7-1.6 0-2.2L6.4 1.4C5.5.5 4 1.2 4 2.5z" />
      </svg>
      <span
        className="visually-hidden">
        January 2017
      </span>
    </button>
  </div>
</div>
`;
